nvflare/app_opt/lightning/fl_callbacks.py

```diff
@@ -0,0 +1,142 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import traceback
+from enum import Enum
+
+import pytorch_lightning as pl
+from pytorch_lightning.trainer.states import TrainerFn
+
+import nvflare.client as flare
+from nvflare.app_common.abstract.fl_model import FLModel
+from nvflare.client.config import ConfigKey
+from nvflare.client.constants import ModelExchangeFormat
+
+
+class SendTrigger(Enum):
+    AFTER_TRAIN_AND_TEST = 1
+    AFTER_TRAIN = 2
+    AFTER_TEST = 3
+
+
+class FLCallback(pl.callbacks.Callback):
+    def __init__(self, send_trigger: SendTrigger = SendTrigger.AFTER_TRAIN_AND_TEST):
+        super(FLCallback, self).__init__()
+        flare.init(
+            config={
+                ConfigKey.EXCHANGE_PATH: "./",
+                ConfigKey.EXCHANGE_FORMAT: ModelExchangeFormat.PYTORCH,
+                ConfigKey.TRANSFER_TYPE: "FULL",
+            }
+        )
+        self.send_mode = send_trigger
+        self.input_fl_model = None
+        self.output_fl_model = None
+        self.metrics_captured = False
+        self.prev_loop_run = None
+
+    def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", stage: str) -> None:
+        loop = trainer.test_loop
+        self.prev_loop_run = loop.run
+        loop.run = test_loop_run_decorator(loop, self)
+
+    def on_fit_start(self, trainer, pl_module):
+        # receive the global model and update the local model with global model
+        # the 1st time test() or fit() is called.
+        self._receive_update_model(pl_module)
+
+    def on_train_end(self, trainer, pl_module):
+        if self.output_fl_model:
+            self.output_fl_model.params = pl_module.cpu().state_dict()
+        else:
+            self.output_fl_model = flare.FLModel(params=pl_module.cpu().state_dict())
```

Comment:
We should call self._check_and_send() here as well?

Otherwise, in your example, 
the first test will receive global model and test on it.
the train will use the cache global model and train on it. (<- we should submit here)
the second test will test on the trained model and in the end submit the trained model.

I think we should not assume the second test will always be there.